### PR TITLE
Feature enemies

### PR DESCRIPTION
**Pull request by Ocsof**

### CODE REVIEW

File: src/main/scala/model/WaveManager.scala [hunk 1]
```diff
@@ -50,4 +50,4 @@ object Generator:
     private def createEnemyList(n: Int)(l: List[Enemy]): List[Enemy] =
       n match
         case 0 => l
-        case _ => createEnemyList(n - 1)(l = l :+ new Zombie((Random.between(0, NumOfLanes), LanesLength + Random.between(0, 20))))
\ No newline at end of file
+        case _ => createEnemyList(n - 1)(l = l :+ FastZombie())
```
Comment (PaoloPenazzi) [inline src/main/scala/model/WaveManager.scala:53]: I suppose this was made for testing purposes.
Revert the changes and make the wave generator spawn only basic zombies.

File: src/main/scala/model/entities/Bullet.scala [hunk 2]
```diff
@@ -47,28 +47,40 @@ abstract class Bullet(position: Position) extends Entity with MovingAbility:
...
+ */
+case class PawBullet(override val position: Position) extends ZombieAttack(position):
+  override def update(elapsedTime: FiniteDuration, interests: List[Entity]): Bullet = this withPosition newPositionAfter(elapsedTime)
+  override def withPosition(pos: Position): Bullet = copy(position = pos)
+
+/**
+ * The [[ZombieAttack]] shoot by  [[WarriorZombie]]. It has no strange effects on the plants beside dealing damage.
+ * @param position The initial position of the [[Bullet]].
+ */
+case class SwordAttack(override val position: Position) extends ZombieAttack(position):
```
Comment (PaoloPenazzi) [inline src/main/scala/model/entities/Bullet.scala:84]: Every entity shoot should be a bullet.
A zombie shoots a SwordBullet not a SwordAttack.
Change the name and use the concept of Bullet instead of the concept of Attack.
The same goes for the abstract class ZombieAttack (should be ZombieBullet)

File: src/main/scala/model/entities/Enemy.scala [hunk 3]
```diff
@@ -41,14 +42,52 @@ abstract class Enemy(override val position: Position,
...
 
 /**
- * Basic Zombie Enemy.
+ * Zombie is the base enemy of the game.
+ *
+ * @param position the position in which the zombie is placed.
+ * @param life the life that the zombie currently has.
+ * @param state the state of the zombie.
+ */
+case class Zombie(override val position: Position = (Random.between(0, NumOfLanes), LanesLength + Random.between(0, 20)),
```
Comment (PaoloPenazzi) [inline src/main/scala/model/entities/Enemy.scala:51]: Consider putting all these magic numbers in the DefaultValues file.

File: src/main/scala/view/GameScreen.scala [hunk 4]
```diff
@@ -134,7 +120,9 @@ class GameScreen() extends ScreenAdapter with EntityRenderer :
 
   def memoizedTexture: Entity => Texture =
 
-    def texture(entity: Entity): Texture = new Texture(Gdx.files.classpath("assets/" + spriteName(entity)))
+    def texture(entity: Entity): Texture =
+      val spritename = spriteName(entity)
```
Comment (PaoloPenazzi) [inline src/main/scala/view/GameScreen.scala:124]: I don't get why you are saving the sprite name and then you're not using it (by the way the val name should be 'spriteName').
I would revert the changes and leave how it was before.

File: src/main/scala/model/entities/Enemy.scala [hunk 5]
```diff
@@ -2,25 +2,26 @@ package model.entities
...
+ * @param life the life that the enemy currently has.
+ * @param state the state of the enemy.
  */
 abstract class Enemy(override val position: Position,
             override val life: Int,
-            override val state: TroopState,
-            override val velocity: Float) extends Troop with MovingAbility:
+            override val state: TroopState) extends Troop with MovingAbility:
 
   override def isInterestedIn: Entity => Boolean =
```
Comment (PaoloPenazzi) [inline src/main/scala/model/entities/Enemy.scala:21]: By looking at this method it's not clear what it does.
Refactor in the same way we did in the Plant file.

File: src/test/scala/model/EnemyModelTest.scala [hunk 6]
```diff
@@ -1,38 +1,90 @@
...
+    List(dummyZombie, plantInTheSameLane, dummyBullet) filter zombie.isInterestedIn shouldBe List.empty
+  }
+
+  "A FastZombie" should "be in Moving state if it can't attack any plant" in {
+    fastZombie.update(FiniteDuration(16, "milliseconds"), List()).state shouldBe Moving
+  }
+  "A FastZombie" should "enter attacking state if it can attack an enemy" in {
+    fastZombie.update(FiniteDuration(16, "milliseconds"), List(plantInTheSameLane)).state shouldBe Attacking
   }
+  "A FastZombie" should "filter the interesting entities" in {
```
Comment (PaoloPenazzi) [inline src/test/scala/model/EnemyModelTest.scala:55]: This feels like a copy paste from the Zombie tests above
Why are you testing again the isInterestIn method? It's the same for every zombie..

File: src/main/scala/model/entities/Enemy.scala [hunk 7]
```diff
@@ -2,25 +2,26 @@ package model.entities
...
 
 /**
- * Interface of an enemy model
+ * A Enemy is an abstract entity that models the common behaviour of different types of enemy.
+ *
+ * @param position the position in which the enemy is placed.
+ * @param life the life that the enemy currently has.
+ * @param state the state of the enemy.
  */
 abstract class Enemy(override val position: Position,
```
Comment (PaoloPenazzi) [inline src/main/scala/model/entities/Enemy.scala:17]: Consider changing the name to Zombie and then find a new name for the basic Zombie.
The game is called Plants vs Zombies so a generic Troop should be a Plant or a Zombie, not a Plant or an Enemy.

File: src/test/scala/model/EnemyModelTest.scala [hunk 8]
```diff
@@ -1,38 +1,90 @@
...
+  private val plantInOtherLane: Troop = Troops.ofType[PeaShooter] withPosition (otherLane, 10)
+  private val plantInTheSameLane: Troop = Troops.ofType[PeaShooter] withPosition (testingLane, 10)
+  private val plantInRange: Troop = Troops.ofType[PeaShooter] withPosition (testingLane, 75)
+
+  private val dummyPlant: Troop = Troops.ofType[PeaShooter] withPosition (otherLane, 50)
+  private val dummyZombie: Troop = Troops.ofType[Zombie] withPosition (otherLane, 50)
+  private val dummyBullet: Bullet = PeaBullet(0,0)
+
+
+  "A Zombie" should "be in Moving state if it can't attack any plant" in {
```
Comment (PaoloPenazzi) [inline src/test/scala/model/EnemyModelTest.scala:30]: More regarding the names: by describing the test as "A zombie should..." it seems that you're testing a generic Zombie.

**Review by PaoloPenazzi — CHANGES_REQUESTED** [review]

File: src/main/scala/view/GameScreen.scala (hunk 4 above)
Comment (Ocsof) [reply src/main/scala/view/GameScreen.scala:124]: Right, just forgot to delete the variable

File: src/main/scala/model/entities/Enemy.scala (hunk 3 above)
Comment (Ocsof) [reply src/main/scala/model/entities/Enemy.scala:51]: yess

**Review by PaoloPenazzi — APPROVED** [review]
